### PR TITLE
Add book titles to search results

### LINKED ISSUE #704: Add book titles to search results
When seaching across multiple zim archives it would be useful to be able to see which archive each search result came from to avoid confusion. 

For instance, in my case I have Wikipedia, archwiki, and python documentation amongst others set up with kiwix-serve. If I seach for something technical like `Python` across all these books, it can be really confusing which book each result came from.

Most public search engines have a bit of text under each search result that shows what site results come from. By adding the book title at the bottom of each result, we can emulate a behaviour lots of users have come to expect.

One possible downside is that the book titles will clutter the search results, but a few words per search result don't seem to distract too much.

### PR DESCRIPTION
**Pull request by thavelick**
Implements/fixes #704 and fixes https://github.com/kiwix/kiwix-tools/issues/478

This is my first pull request here, so appologies if I've missed any process or coding style guidlines or best practices. Happy to revise anything or provide additional context.

It would be good as well to have a corresponding PR in kiwix/kiwix-tools and kiwix/kiwix-desktop to see how it works from a Kiwix port / end-user perspective.

### CODE REVIEW

**Comment by thavelick** [comment]
I realized after the fact that by changing the constructor of `SearchRenderer`, this may break 3rd party software that uses libkiwix. I could see either:

1. Adding a new constructor that adds the libary but keeping the original one OR
2. Refactoring  this to set the library with a setter instead of with a constructor 

I happy to make one of the above changes but I want to make sure there's enough interest in this enhancement being merged in to the main rpeo before I spend a lot of time making it perfect. It works for my own purposes as is.

**Comment by kelson42** [comment]
@thavelick Thank you for your first PR. We will review it and give you a feedback soon.

File: src/search_renderer.cpp [hunk 1]
```diff
@@ -90,6 +93,11 @@ std::string SearchRenderer::getHtml()
     std::ostringstream s;
     s << it.getZimId();
     result.set("resultContentId", mp_nameMapper->getNameForId(s.str()));
+    std::shared_ptr<zim::Archive> archive;
+    try {
+       archive = mp_library->getArchiveById(s.str());
+       result.set("bookTitle", archive ? getArchiveTitle(*archive) : "");
```
Comment (veloman-yunkan) [inline src/search_renderer.cpp:99]: An empty string for `bookTitle` won't disable the mustache block marked with `{{#bookTitle}}...{{/bookTitle}}`. You will need to pass a `kainjow::mustache::data(false)` instead.

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]
The two commits should be combined into one.

Regarding the non-backward-compatible change in the public API of `SearchRenderer`, I personally wouldn't mind about it, but the final word should belong to @mgautierfr.

**Comment by kelson42** [comment]
@thavelick Any fix in view regarding @veloman-yunkan feedbacks?

**Comment by thavelick** [comment]
@kelson42 I've been busy with work stuff and haven't had a chance to look the feedback in any depth yet. I think I should be able to address everything here this weekend.

File: src/search_renderer.cpp (hunk 1 above)
Comment (thavelick) [reply src/search_renderer.cpp]: Updated that so it uses mustache's false.  Really this should be pertty much impossible. Both the `{#bookTitle}` wrapper and this null check are meant as guards against future bugs.

**Comment by thavelick** [comment]
@veloman-yunkan @kelson42 I've squashed my commits to one and addressed your comment

@mgautierfr looking forward to hearing your thoughts on the constructor.

**Comment by thavelick** [comment]
Note, the ability to have mulitlpe seach results is currently a little hidden I've been testing this by starting up `kiwix-serve` and then hitting `/search?pattern=emulator` which will trigger resutls across archives if you have both wikipedia and wikihow zimfiles.

**Comment by kelson42** [comment]
@thavelick Thx for the update. It is true that the multizim search feature is not proeminent. This is because it does not scale in appropriate manner. This is a problem @maneeshpm is working on and I hope the problem will be fixed in one of the next minor releases.

File: src/search_renderer.cpp [hunk 2]
```diff
@@ -90,6 +93,11 @@ std::string SearchRenderer::getHtml()
     std::ostringstream s;
     s << it.getZimId();
     result.set("resultContentId", mp_nameMapper->getNameForId(s.str()));
+    std::shared_ptr<zim::Archive> archive;
+    try {
+       archive = mp_library->getArchiveById(s.str());
+       result.set("bookTitle", archive ? getArchiveTitle(*archive) : kainjow::mustache::data(false));
```
[the remark this answers is not in the document]
Comment (thavelick) [reply src/search_renderer.cpp]: @veloman-yunkan Your right, I don't think there is a situation in which archive can be null here. I've switched to just using `mp_library->getBookById(s.str()).getTitle()` as you've suggested

**Comment by thavelick** [comment]
@veloman-yunkan @kelson42 Let me know if you need anything else form me to get this one over the line

**Review by veloman-yunkan — APPROVED** [review]

**Pull request merged**